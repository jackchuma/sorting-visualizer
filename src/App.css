html {
  text-align: center;
  color: hsla(202, 100%, 10%, 1);
  background-color: hsla(202, 21%, 80%, 1);
}

.button:hover {
  background-color: hsla(202, 30%, 80%, 1);
}

h1 {
  color: hsla(202, 60%, 15%, 1)
}

.hidden-nav {
  display: flex;
  flex-direction: column;
  justify-content: top;
  align-items: center;
  position: fixed;
  height: 100%;
  width: 200px;
  background-color: hsla(202, 100%, 10%, 1);
  color: white;
  padding-top: 50px;
  border: 1px solid hsla(202, 100%, 10%, 1);
  left: -200px;
  transition: left 1s;
}

.hidden-nav-button {
  height: 30px;
  background-color:hsla(202, 100%, 10%, 1);
  color: white;
  border: none;
  width: 100%;
}

.hidden-nav-button:hover {
  cursor: pointer;
  background-color: hsla(202, 21%, 80%, 1);
}

.hidden-nav-button:active {
  color:hsla(202, 100%, 10%, 1);
}

.hidden-sort-list {
  display: none;
  flex-direction: column;
  width: 100%;
}

.hidden-arrow {
  gap: 5px;
  display: flex;
  align-items: center;
  justify-content: center;
}

#hidden-slider {
  display: none;
}

.navbar {
  background-color: hsla(202, 100%, 10%, 1);
  height: 50px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.navbar-left {
  display: none;
}

.navbar-right {
  display: flex;
  justify-content: center;
  align-items: center;
}

.hamburger {
  background-color: hsla(202, 100%, 10%, 1);
  color: white;
  border: none;
  display: none;
  left: 0;
  transform: scale(1.8);
}

.array-button {
  margin-right: 100px;
  background-color: hsla(202, 60%, 80%, 1);
  border: none;
  height: 30px;
  border-radius: 15px;
  padding: 0 20px;
}

.array-button:hover {
  cursor: pointer;
  transform: scale(1.1);
}

.array-button:active {
  transform: scale(1);
}

.array-size {
  background-color: hsla(202, 21%, 80%, 1);
  display: flex;
  align-items: center;
  flex-direction: column;
  justify-content: top;
  height: 30px;
  border-radius: 15px;
}

.array-size-button {
  background-color: hsla(202, 60%, 80%, 1);
  height: 30px;
  border: none;
  display: flex;
  align-items: center;
  padding: 0 20px;
  border-radius: 15px;
  gap: 5px;
}

.array-size-button:hover {
  cursor: pointer;
}

.array-slider {
  border: 1px solid hsla(202, 21%, 50%, 1);
  display: none;
  background-color: hsla(202, 21%, 80%, 1);
  z-index: 2;
  border-radius: 4px;
  box-shadow: 0 2px 4px 0 rgba(28, 28, 28, 0.2);
  position: absolute;
  margin-top: 30px;
}

.slider {
  padding: 10px 10px;
  background-color: hsla(202, 60%, 30%, 1);
}

.slider::-webkit-slider-thumb {
  cursor: pointer;
}

.sort-button {
  background-color: hsla(34, 100%, 50%, 1);
  padding: 10px 20px;
  border-radius: 3px;
  margin-left: 100px;
  margin-right: 100px;
  border: none;
  font-weight: 700;
  font-size: 16px;
}

.sort-button:hover {
  transform: scale(1.1);
  cursor: pointer;
}

.sort-button:active {
  transform: scale(1);
}

.sort-menu {
  background-color: hsla(202, 21%, 80%, 1);
  display: flex;
  align-items: center;
  flex-direction: column;
  justify-content: top;
  height: 30px;
  border-radius: 15px;
}

.sort-button-visible {
  background-color: hsla(202, 60%, 80%, 1);
  min-height: 30px;
  border: none;
  display: flex;
  align-items: center;
  padding: 10px 20px;
  border-radius: 15px;
  gap: 5px;
}

.sort-button-visible:hover {
  cursor: pointer;
}

.sort-list {
  border: 1px solid hsla(202, 21%, 50%, 1);
  display: none;
  flex-direction: column;
  background-color: hsla(202, 21%, 80%, 1);
  z-index: 2;
  border-radius: 4px;
  box-shadow: 0 2px 4px 0 rgba(28, 28, 28, 0.2);
  position: absolute;
  margin-top: 30px;
  overflow: hidden;
}

.sort-list button {
  background-color: hsla(202, 21%, 80%, 1);
  text-align: left;
  border: none;
  padding: 5px 20px;
  letter-spacing: 0.5px;
}

.sort-list button.selected,
.sort-list button:hover {
  color: hsla(34, 80%, 50%, 1);
  background-color: hsla(202, 21%, 50%, 1);
  cursor: pointer;
}

.sorting-view-container {
  display: flex;
  justify-content: center;
}

.sorting-view {
  background-color: hsla(202, 21%, 50%, 1);
  height: 500px;
  width: 1000px;
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  border: 3px solid hsla(202, 100%, 10%, 1);
}

.rectangle {
  background-color: hsla(34, 80%, 50%, 1);
}

@media only screen and (max-width: 1000px) {
  .sorting-view {
    width: 75%;
  }
}

@media only screen and (max-width: 850px) {
  .navbar-left {
    display: flex;
    align-items: center;
    height: 100%;
    width: 100%;
  }

  .navbar-right {
    display: flex;
    align-items: center;
    height: 100%;
    width: 100%;
    justify-content: right;
  }

  .rectangle {
    transform: scaleX(0.5);
  }

  .sort-button {
    margin-right: 15%;
  }

  .array-button {
    display: none;
  }

  .array-size {
    display: none;
  }

  .sort-menu {
    display: none;
  }

  .hamburger {
    display: flex;
    align-items: center;
    justify-content: center;
    margin-left: 30px;
  }

  .hamburger:hover {
    background-color: hsla(202, 30%, 50%, 1);
    cursor: pointer;
  }
}