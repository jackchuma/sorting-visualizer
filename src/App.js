import React, { useState } from 'react';
import './App.css';
import { IoIosArrowDown } from 'react-icons/io';

function App() {

  const [selectedSort, setSelectedSort] = useState({name: 'Bubble Sort', id: 'bubble-sort'});
  const [sliderVal, setSliderVal] = useState(50);
  const [array, setArray] = useState([]);

  const generateArray = () => {
    let arr = [];
    for (let i=0; i<sliderVal; i++) {
      arr.push(Math.floor(Math.random() * 450 + 50));
    }
    setArray(arr);
  }

  const handleArraySizeButtonClick = () => {
    const sliderBox = document.getElementById('array-slider');

    if (sliderBox.style.display === 'block') {
      sliderBox.style.display = 'none';
    } else {
      sliderBox.style.display = 'block';
    }
  }

  const handleSortButtonClick = e => {
    const sortBox = document.getElementById('sort-list');

    if (sortBox.style.display === 'flex') {
      sortBox.style.display = 'none';
    } else {
      sortBox.style.display = 'flex';
    }
  }

  const moveSlider = e => {
    setSliderVal(e.target.value);
    generateArray();
  }

  const handleSortListButtonClick = e => {
    const oldLink = document.getElementById(selectedSort.id);
    oldLink.style.backgroundColor = 'hsla(202, 21%, 80%, 1)';
    oldLink.style.color = 'black';
    setSelectedSort({name: e.target.value, id: e.target.id});

    const newLink = document.getElementById(e.target.id);
    newLink.style.color = 'hsla(34, 80%, 50%, 1)';
    newLink.style.backgroundColor = 'hsla(202, 21%, 50%, 1)';

    const sortBox = document.getElementById('sort-list');
    sortBox.style.display = 'none';
  }

  const bubbleSort = async input => {
    let swapping = true;
    while (swapping) {
      swapping = false;
      for (let i=0; i<input.length-1; i++) {
        let first = document.getElementById(i);
        let next = document.getElementById(i+1);

        if (input[i] > input[i+1]) {
          first.style.backgroundColor = 'green';
          next.style.backgroundColor = 'red';
          swap(input, i, i+1);
          await sleep(5);
          first.style.order = `${i}`;
          first.id = i+1;
          next.style.order = `${i-1}`;
          next.id = i;
          await sleep(5);
          swapping = true;
        } else {
          first.style.backgroundColor = 'red';
          next.style.backgroundColor = 'green';
          await sleep(5);
        }
        
        first.style.backgroundColor = 'hsla(34, 80%, 50%, 1)';
        next.style.backgroundColor = 'hsla(34, 80%, 50%, 1)';
      }
      setArray(input);
    }
    return input;
  }

  const mergeSort = input => {
    const length = input.length;
    if (length === 1) {
      return input;
    }
    const mid = Math.floor(length / 2);
    const leftArray = input.slice(0, mid);
    const rightArray = input.slice(mid, length);
    
    return merge(mergeSort(leftArray), mergeSort(rightArray));
  }

  const merge = (leftArray, rightArray) => {
    const sortedArray = [];
    while (leftArray.length > 0 && rightArray.length > 0) {
      
      if (leftArray[0] < rightArray[0]) {
        sortedArray.push(leftArray.shift());
      } else {
        sortedArray.push(rightArray.shift());
      }
    }
    setArray(sortedArray.concat(leftArray).concat(rightArray));
    return sortedArray.concat(leftArray).concat(rightArray);
  }

  const swap = (arr, indexOne, indexTwo) => {
    const temp = arr[indexTwo];
    arr[indexTwo] = arr[indexOne];
    arr[indexOne] = temp;
  }

  const handleSort = () => {
    if (selectedSort.name === 'Bubble Sort') {
      bubbleSort(array);
    } else if (selectedSort.name === 'Merge Sort') {
      mergeSort(array);
    }
  }

  const sleep = time => {
    return new Promise(resolve => setTimeout(resolve, time));
  }

  const sortList = [{name: 'Bubble Sort', id: 'bubble-sort'}, {name: 'Quick Sort', id: 'quick-sort'}, {name: 'Merge Sort', id: 'merge-sort'}];

  return (
    <div className="App">
      <div className="navbar">
        <button className='array-button' onClick={generateArray}>Generate New Array</button>
        <div className='array-size'>
          <button className='array-size-button' onClick={handleArraySizeButtonClick}>Change Array Size <IoIosArrowDown /></button>
          <div className='array-slider' id='array-slider'>
            <input type='range' className='slider' min='10' onChange={moveSlider} />
          </div>
        </div>
        <button className="sort-button" onClick={handleSort}>Sort!</button>
        <div className="sort-menu" id="sort-menu">
          <button className='sort-button-visible' onClick={handleSortButtonClick}>{selectedSort.name} <IoIosArrowDown /></button>
          <div className='sort-list' id='sort-list'>
            {sortList.map((sortName) => {
              return <button key={sortName.id} id={sortName.id} value={sortName.name} onClick={handleSortListButtonClick}>{sortName.name}</button>
            })}
          </div>
        </div>
      </div>
      <h1>Jack's Sorting Algorithm Visualizer</h1>
      <div className='sorting-view-container'>
        <div className='sorting-view'>
          {array.map((val, index) => {
            return <div className='rectangle' key={index} id={index} style={{height: val + 'px', width: (1000 / array.length - 5) + 'px', order: index+1}}></div>
          })}
        </div>
      </div>
    </div>
  );
}

export default App;
