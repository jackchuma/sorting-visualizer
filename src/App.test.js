import React from 'react';
import App from './App';
import renderer from 'react-test-renderer';

test('App renders without crashing', () => {
  const app = renderer.create(<App />);
  let tree = app.toJSON();
  expect(tree).toMatchSnapshot();
})
