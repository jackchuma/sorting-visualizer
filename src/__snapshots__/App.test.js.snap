// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing 1`] = `
<div
  className="App"
>
  <div
    className="hidden-nav"
    id="hidden-nav"
  >
    <button
      className="hidden-nav-button"
      onClick={[Function]}
    >
      Generate New Array
    </button>
    <button
      className="hidden-nav-button hidden-arrow"
      onClick={[Function]}
    >
      Change Array Size 
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
        />
      </svg>
    </button>
    <input
      className="hidden-nav-button hidden-nav-slider"
      id="hidden-slider"
      min="10"
      onChange={[Function]}
      type="range"
    />
    <button
      className="hidden-nav-button hidden-arrow"
      onClick={[Function]}
    >
      Bubble Sort
       
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
        />
      </svg>
    </button>
    <div
      className="hidden-sort-list"
      id="hidden-sort-list"
    >
      <button
        className="hidden-nav-button"
        onClick={[Function]}
        value="Bubble Sort"
      >
        Bubble Sort
      </button>
      <button
        className="hidden-nav-button"
        onClick={[Function]}
        value="Quick Sort"
      >
        Quick Sort
      </button>
      <button
        className="hidden-nav-button"
        onClick={[Function]}
        value="Merge Sort"
      >
        Merge Sort
      </button>
    </div>
  </div>
  <div
    className="navbar"
  >
    <div
      className="navbar-left"
    >
      <button
        className="hamburger"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 6h16v2H4zm0 5h16v2H4zm0 5h16v2H4z"
          />
        </svg>
      </button>
    </div>
    <div
      className="navbar-right"
    >
      <button
        className="array-button button"
        onClick={[Function]}
      >
        Generate New Array
      </button>
      <div
        className="array-size"
      >
        <button
          className="array-size-button button"
          onClick={[Function]}
        >
          Change Array Size 
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
            />
          </svg>
        </button>
        <div
          className="array-slider"
          id="array-slider"
        >
          <input
            className="slider"
            min="10"
            onChange={[Function]}
            type="range"
          />
        </div>
      </div>
      <button
        className="sort-button"
        onClick={[Function]}
      >
        Sort!
      </button>
      <div
        className="sort-menu"
        id="sort-menu"
      >
        <button
          className="sort-button-visible button"
          onClick={[Function]}
        >
          Bubble Sort
           
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
            />
          </svg>
        </button>
        <div
          className="sort-list"
          id="sort-list"
        >
          <button
            id="bubble-sort"
            onClick={[Function]}
            value="Bubble Sort"
          >
            Bubble Sort
          </button>
          <button
            id="quick-sort"
            onClick={[Function]}
            value="Quick Sort"
          >
            Quick Sort
          </button>
          <button
            id="merge-sort"
            onClick={[Function]}
            value="Merge Sort"
          >
            Merge Sort
          </button>
        </div>
      </div>
    </div>
  </div>
  <h1>
    Jack's Sorting Algorithm Visualizer
  </h1>
  <div
    className="sorting-view-container"
  >
    <div
      className="sorting-view"
      id="sorting-view"
    />
  </div>
</div>
`;
